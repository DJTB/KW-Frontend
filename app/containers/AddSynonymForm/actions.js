
import {
  LOAD_JISHODATA,
  LOAD_JISHODATA_SUCCESS,
  LOAD_JISHODATA_ERROR,
  ADD_SYNONYM,
  ADD_SYNONYM_ERROR,
} from './constants';

export function loadJishoData(keyword) {
  return {
    type: LOAD_JISHODATA,
    payload: keyword,
  };
}

export function jishoDataLoaded(data) {
  return {
    type: LOAD_JISHODATA_SUCCESS,
    payload: data,
  };
}

export function jishoDataLoadingError(error) {
  return {
    type: LOAD_JISHODATA_ERROR,
    payload: error,
  };
}

export function addSynonym(data) {
  return {
    type: ADD_SYNONYM,
    payload: data,
  };
}


export function addSynonymError(error) {
  return {
    type: ADD_SYNONYM_ERROR,
    payload: error,
  };
}
