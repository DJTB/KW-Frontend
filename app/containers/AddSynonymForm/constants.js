 export const LOAD_JISHODATA = 'kw/AddSynonymForm/LOAD_JISHODATA';
 export const LOAD_JISHODATA_SUCCESS = 'kw/AddSynonymForm/LOAD_JISHODATA_SUCCESS';
 export const LOAD_JISHODATA_ERROR = 'kw/AddSynonymForm/LOAD_JISHODATA_ERROR';

 export const ADD_SYNONYM = 'kw/AddSynonymForm/ADD_SYNONYM';
 export const ADD_SYNONYM_SUCCESS = 'kw/AddSynonymForm/ADD_SYNONYM_SUCCESS';
 export const ADD_SYNONYM_ERROR = 'kw/AddSynonymForm/ADD_SYNONYM_ERROR';
 export const REMOVE_SYNONYM = 'kw/AddSynonymForm/REMOVE_SYNONYM';
 export const REMOVE_SYNONYM_ERROR = 'kw/AddSynonymForm/REMOVE_SYNONYM_ERROR';
