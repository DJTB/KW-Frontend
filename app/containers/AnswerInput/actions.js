/*
 *
 * AnswerInput actions
 *
 */

import {
  UPDATE_INPUT,
} from './constants';

export function updateInput(data) { // eslint-disable-line import/prefer-default-export
  return {
    type: UPDATE_INPUT,
    data,
  };
}
