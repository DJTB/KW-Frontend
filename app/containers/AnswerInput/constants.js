/*
 *
 * AnswerInput constants
 *
 */

export const INPUT_CHANGED = 'kw/AnswerInput/INPUT_CHANGED';
export const UPDATE_INPUT = 'kw/AnswerInput/UPDATE_INPUT';
