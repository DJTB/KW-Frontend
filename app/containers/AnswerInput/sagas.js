import { call } from 'redux-saga/effects';

// Individual exports for testing
export function* answerInputSaga() {
  yield call(console.log, 'test'); // eslint-disable-line no-console
}

// All sagas to be loaded
export default [
  answerInputSaga,
];
