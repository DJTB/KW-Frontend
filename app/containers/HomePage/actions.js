import actionTypeCreator, { /* ASYNC*/ } from 'redux-action-types-creator';
// import { createAction } from 'redux-actions';

export const TYPES = actionTypeCreator('APP/HOMEPAGE', { asyncSuffix: ['REQUEST', 'SUCCESS', 'FAILURE', 'CANCEL'] })({
//  USER: ASYNC, // request, success, failure, cancel
});
//
// export const userLoad = createAction(TYPES.USER.LOAD, (payload) => payload, (payload, meta) => meta);
// export const userLoadSuccess = createAction(TYPES.USER.SUCCESS); // auto identity for payload: action(payloadVal)
// export const userLoadFailure = createAction(TYPES.USER.FAILURE);
// export const userLoadCancel = createAction(TYPES.USER.CANCEL);
