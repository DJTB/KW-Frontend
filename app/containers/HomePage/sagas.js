import { take, call, put, fork, cancel } from 'redux-saga/effects'; // eslint-disable-line no-unused-vars
import { LOCATION_CHANGE } from 'react-router-redux';

/**
 * Root saga manages watcher lifecycle
 */
export function* homeSaga() {
  // Fork watcher so we can continue execution
  const watcher = yield fork(/* watchers */);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [];
