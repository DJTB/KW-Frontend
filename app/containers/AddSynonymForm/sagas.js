import { takeEvery, put, call, select } from 'redux-saga/effects';
import { createSynonymUrl } from 'shared/urls';
import markAllAsDaemon from 'utils/markAllAsDaemon';
import post from 'utils/post';
// import request from 'utils/request';
// import { createJishoUrl } from 'shared/urls';
// import shapeJishoData from './utils/shapeJishoData';
// import { LOAD_JISHODATA } from './constants';
import { ADD_SYNONYM, REMOVE_SYNONYM } from './constants';

import { addSynonymToCurrent, removeSynonymFromCurrent } from 'containers/ReviewSession/actions';
import { selectCurrent } from 'containers/ReviewSession/selectors';
import { checkAnswer } from 'containers/ReviewAnswer/actions';

import {
  addSynonymError,
  addSynonymSuccess,
  removeSynonymError,
  // jishoDataLoaded,
  // jishoDataLoadingError,
} from './actions';

// export function* getJishoData({ payload }) {
//   const requestURL = createJishoUrl(payload);
//   const requestOptions = { mode: 'no-cors' };
//   try {
//     const data = yield call(request, requestURL, requestOptions);
//     const shapedData = shapeJishoData(data);
//     yield put(jishoDataLoaded(shapedData));
//   } catch (err) {
//     yield put(jishoDataLoadingError(err));
//   }
// }

export function* postNewSynonym({ payload }) {
  const current = yield select(selectCurrent());
  const synonym = {
    character: payload.get('Kanji'),
    kana: payload.get('Kana'),
  };
  const postUrl = createSynonymUrl();
  const postData = {
    review: current.get('id'),
    ...synonym,
  };

  try {
    const synonymData = yield call(post, postUrl, postData);
    yield put(addSynonymSuccess({
      title: 'Add Synonym',
      message: `Synonym “${synonymData.character}” successfully added.`,
    }));
    yield put(addSynonymToCurrent(synonymData));
    yield put(checkAnswer());
  } catch (err) {
    console.error(err);
    yield put(addSynonymError({
      title: 'Connection error',
      message: `Unable to add synonym on server: ${err.message}`,
    }));
  }
}

// TODO: not really relevant to synonymForm at all
// move to app level since vocab will use it?
// same goes for addSynonym really
export function* removeSynonym({ payload }) {
  const id = payload;
  const url = createSynonymUrl(id);

  try {
    yield call(post, url, null, { method: 'DELETE' });
    yield put(removeSynonymFromCurrent(id));
  } catch (err) {
    console.error(err);
    yield put(removeSynonymError({
      title: 'Connection error',
      message: `Unable to remove synonym on server: ${err.message}`,
    }));
  }
}


// export function* getJishoDataWatcher() {
//   yield takeLatest(LOAD_JISHODATA, getJishoData);
// }

export function* addSynonymWatcher() {
  yield takeEvery(ADD_SYNONYM, postNewSynonym);
}
export function* removeSynonymWatcher() {
  yield takeEvery(REMOVE_SYNONYM, removeSynonym);
}

// Mark watchers to only run once on route entry
const watchers = markAllAsDaemon([
  addSynonymWatcher,
  removeSynonymWatcher,
/* getJishoDataWatcher, */
]);

// Bootstrap sagas
export default [
  ...watchers,
];
