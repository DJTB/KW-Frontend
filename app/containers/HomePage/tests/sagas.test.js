/**
 * Tests for HomePage sagas
 */

import expect from 'expect';
import { takeLatest } from 'redux-saga';
import { take, call, put, fork, cancel } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';

import { getUserData, getUserDataWatcher, userData } from '../sagas';

import { LOAD_USERDATA } from 'containers/App/constants';
import { userDataLoaded, userDataLoadingError } from 'containers/App/actions';

import request from 'utils/request';

describe('getUserData Saga', () => {
  let getUserDataGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getUserDataGenerator = getUserData();

    const requestURL = 'api/profiles';
    const callDescriptor = getUserDataGenerator.next().value;
    expect(callDescriptor).toEqual(call(request, requestURL));
  });

  it('should dispatch the userDataLoaded action if it requests the data successfully', () => {
    const response = {
      name: 'userData name',
    };
    const putDescriptor = getUserDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(userDataLoaded(response)));
  });

  it('should call the userDataLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getUserDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(userDataLoadingError(response)));
  });
});

describe('getUserDataWatcher Saga', () => {
  const getUserDataWatcherGenerator = getUserDataWatcher();

  it('should watch for LOAD_USERDATA action', () => {
    const takeDescriptor = getUserDataWatcherGenerator.next().value;
    expect(takeDescriptor).toEqual(fork(takeLatest, LOAD_USERDATA, getUserData));
  });
});

describe('UserDataSaga Saga', () => {
  const UserDataSaga = userData();

  let forkDescriptor;

  it('should asyncronously fork getUserDataWatcher saga', () => {
    forkDescriptor = UserDataSaga.next();
    expect(forkDescriptor.value).toEqual(fork(getUserDataWatcher));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = UserDataSaga.next();
    expect(takeDescriptor.value).toEqual(take(LOCATION_CHANGE));
  });

  it('should finally cancel() the forked getUserDataWatcher saga',
     function* UserDataSagaCancellable() {
      // reuse open fork for more integrated approach
       forkDescriptor = UserDataSaga.next(put(LOCATION_CHANGE));
       expect(forkDescriptor.value).toEqual(cancel(forkDescriptor));
     }
   );
});
