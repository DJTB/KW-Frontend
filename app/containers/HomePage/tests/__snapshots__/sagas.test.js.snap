exports[`getUserData Saga should call the userDataLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "api/profiles",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getUserData Saga should call the userDataLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "error": [TypeError: Cannot read property 'name' of undefined],
      "type": "kw/App/LOAD_USERDATA_ERROR",
    },
    "channel": null,
  },
}
`;

exports[`getUserData Saga should dispatch the userDataLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "api/profiles",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getUserData Saga should dispatch the userDataLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "error": [TypeError: Cannot read property 'name' of undefined],
      "type": "kw/App/LOAD_USERDATA_ERROR",
    },
    "channel": null,
  },
}
`;
